use source::encoding

khwilliamson opened the pull request:
I'm having trouble thinking of diabolical tests for this, so please try some out on this

khwilliamson committed:
Draft: add use source::encoding

leonerd commented on toke.c:
```diff
@@ -775,6 +777,42 @@ S_cr_textfilter(pTHX_ int idx, SV *sv, int maxlen)
+{
+    PERL_ARGS_ASSERT_YYERROR_NON_ASCII_MESSAGE;
+
+    yyerror_pv(Perl_form(aTHX_ "Use of non-ASCII character 0x%02X"
+                               " illegal when 'use source::encoding"
+                               " \"ASCII\"' is in effect", *s), 0);
```
I'd suggest putting a linebreak in after the `aTHX_` so you can start the string literal much closer to the left, and hence use two longer lines rather than these three tightly-wrapped ones.

khwilliamson replied:
ok

leonerd commented on pod/perldiag.pod:
```diff
@@ -7357,6 +7357,10 @@ See L<perlunicode/Wildcards in Property Values>.
 it already went past any symlink you are presumably trying to look for.
 The operation returned C<undef>.  Use a filename instead.
 
+=item Use of non-ASCII character 0x%X illegal when 'use source::encoding "ASCII"' is in effect
```
The pragma module spells it `"ascii"` in lower-case. I don't have any strong opinion on which case it should be, but this message ought to agree with the module itself.

khwilliamson replied:
I didn't mention that case doesn't matter.  I think that's the best option.  And I could mention it, or not; but if I don't, then it ought to be consistently cased, yes

Grinnz replied:
I think case should matter. Accepting extra ways to specify the same thing just makes the interface more complex and increases future design debt.

khwilliamson replied:
change to make lower case required

Leont commented on lib/source/encoding.pm:
```diff
@@ -0,0 +1,77 @@
+    utf8->unimport;
+}
+
+our $AUTOLOAD;
```
Why this autoload here?

khwilliamson replied:
cargo culting

khwilliamson replied:
removed

khwilliamson committed:
toke.c: Split a variable into two for clarity

It can have two different meanings; split and rename to clarify what
meaning it refers to.

khwilliamson committed:
toke.c: Rename some variables; terminology in comment

Previously in places, things were called variously delimiter and
terminator, or variants thereof.  This makes things consistent, and
clearer.

khwilliamson committed:
toke.c: Rmv unnecessary conditionals

Cheaper to just redo a simple assignment than to test if you've already
done it and skipping it if you had.

khwilliamson committed:
toke.c: white-space only

This outdents some code that the next commit will remove an enclosing
block from

khwilliamson committed:
toke.c: merge loops, multi-byte delim

The code in toke.c had two closely related loops; one for unmirrored
delimiters (same on both ends of the string) that could take UTF-8
delimiters, and one that allowed a mirrored closing delimiter, which
could take only a single byte.  I found that it was just easiest to
collapse these into one loop in preparation to allow multi-byte
mirroring.

rjbs committed:
source::encoding: begin adding tests

khwilliamson committed:
toke.c: Rmv unnecessary SV

This code created an SV simply to get SVfARG() to print it out.  (SVfARG
seems buggy, and I will look at it and fix or file a ticket if so.)  But
nowadays there is UTF8fARG which does the right thing on strings, so the
SV is unnecessary.

The code also has a fix where non-ASCII but Latin1 delimiters  when 'use
utf8' is in effect would potentially come out as garbage  This commit
also adds a clause in a conditional to prevent that.

khwilliamson commented:
Other proposals have made this obsolete, so am closing it
